src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs

```diff
@@ -12,216 +13,194 @@ namespace System
+                    index = str.AsSpan(start, index).IndexOfAnyInRange('A', 'Z');
                 }
-                if (str[i] == ':')
-                    end = i;
             }
 
-            res ??= str.Substring(start, end - start);
+            Debug.Assert(index == -1 || char.IsAsciiLetterUpper(str[start + index]));
+
+            if (index >= 0)
+            {
+                // We saw uppercase letters. Avoid allocating both the substring and the lower-cased variant.
+                return string.Create(end - start, (str, start), static (buffer, state) =>
+                {
+                    int newLength = state.str.AsSpan(state.start, buffer.Length).ToLowerInvariant(buffer);
+                    Debug.Assert(newLength == buffer.Length);
+                });
+            }
+
+            string res = str.Substring(start, end - start);
 
-            if (res == Localhost || res == Loopback)
+            if (res is Localhost or Loopback)
             {
                 loopback = true;
                 return Localhost;
             }
+
             return res;
         }
 
-        //
-        // IsValid
-        //
-        //  Determines whether a string is a valid domain name
-        //
-        //      subdomain -> <label> | <label> "." <subdomain>
-        //
-        // Inputs:
-        //  - name as Name to test
-        //  - starting position
-        //  - ending position
-        //
-        // Outputs:
-        //  The end position of a valid domain name string, the canonical flag if found so
-        //
-        // Returns:
-        //  bool
-        //
-        //  Remarks: Optimized for speed as a most common case,
-        //           MUST NOT be used unless all input indexes are verified and trusted.
-        //
-
-        internal static unsafe bool IsValid(char* name, int pos, ref int returnedEnd, ref bool notCanonical, bool notImplicitFile)
+        public static bool IsValid(ReadOnlySpan<char> hostname, bool iri, bool notImplicitFile, out int length)
         {
-            char* curPos = name + pos;
-            char* newPos = curPos;
-            char* end = name + returnedEnd;
-            for (; newPos < end; ++newPos)
+            int invalidCharOrDelimiterIndex = iri
+                ? hostname.IndexOfAny(s_iriInvalidAsciiChars)
+                : hostname.IndexOfAnyExcept(s_validChars);
+
+            if (invalidCharOrDelimiterIndex >= 0)
             {
-                char ch = *newPos;
-                if (ch > 0x7f) return false;    // not ascii
-                if (ch < 'a') // Optimize for lower-case letters, which make up the majority of most Uris, and which are all greater than symbols checked for below
+                char c = hostname[invalidCharOrDelimiterIndex];
+
+                if (c == '/' || c == '\\' || (notImplicitFile && (c == ':' || c == '?' || c == '#')))
```

Comment:
Nit: as long as you're changing this, you could use newer syntax:
```suggestion
                if (c is '/' or '\\' || (notImplicitFile && (c is ':' or '?' or '#')))
```